packages/suite/src/components/suite/NumberInput.tsx

```diff
@@ -0,0 +1,227 @@
+
+        const { thousandsSeparatorCharCode } = getLocaleSeparatorCharCodes(locale);
+        // allow inputs like '.031' or ',1' and disallow anything non-numerical
+        if (
+            !DECIMAL_SEPARATOR_CODES.includes(inputValue.charCodeAt(0)) &&
+            Number.isNaN(Number(cleanInput))
+        ) {
+            return;
+        }
+
+        // read cursor position before formatting
+        let { selectionStart } = ref.current;
+
+        if (inputValue.length < value.length) {
+            // handle deleting a thousands separator for a non-empty input,
+            // otherwise the number instantly gets formatted back to having that separator
+            const deletedCharacterCode = displayValue[selectionStart]?.charCodeAt(0);
+            if (deletedCharacterCode === thousandsSeparatorCharCode && inputValue.length) {
+                inputValue =
+                    displayValue.substring(0, selectionStart - 1) +
+                    displayValue.substring(selectionStart);
+            }
+        }
+
+        // format and set display value
+        const currentValueLength = ref.current.value.length;
+        const formattedValue = formatDisplayValue(inputValue);
+        if (!formattedValue) {
+            return;
+        }
+        const formattedValueLength = formattedValue.length;
+
+        // pass cleaned value to the form
+        previousFormValue.current = cleanInput;
+        onChange(cleanInput);
+        onChangeCallback?.(cleanInput);
+
+        // detect if spaces have been added/removed
+        if (currentValueLength - formattedValueLength < 0) {
+            selectionStart++;
+        } else if (currentValueLength - formattedValueLength > 0) {
+            selectionStart--;
+        }
+
+        // make sure the cursor stays in the right place after formatting
+        ref.current.setSelectionRange(selectionStart, selectionStart);
+    };
+
+    // copy the non-formatted value
+    const handleCopy = useCallback(
+        (e: React.ClipboardEvent<HTMLInputElement>) => {
+            e.clipboardData.setData('text/plain', cleanValueString(displayValue, locale));
+            e.preventDefault();
+        },
+        [displayValue, locale],
+    );
+
+    const handleKeyDown = useCallback((e: React.KeyboardEvent<HTMLInputElement>) => {
+        // Allow number keys 0-9
+        if (Number(e.key) >= 0 && Number(e.key) <= 9) return;
+        // Allow backspace and delete keys
+        if (e.key === 'Backspace' || e.key === 'Delete') return;
+        // Allow separators
+        if (e.key === ',' || e.key === '.') return;
+        // Allow navigation
+        if (e.key === 'ArrowLeft' || e.key === 'ArrowRight' || e.key === 'Tab') return;
+        // Allow selection, copying and pasting
+        if ((e.ctrlKey || e.metaKey) && ['KeyA', 'KeyC', 'KeyV'].includes(e.code)) return;
+        // Prevent any other keys from being pressed
+        e.preventDefault();
+    }, []);
```

Comment:
Have you tried `onBeforeInput` instead of `onKeyDown`? This only cares about inputting characters, so no need to handle arrows, backspace etc. I tried this (not too sure about the type nor the regex):

```suggestion
    const handleBeforeInput: React.CompositionEventHandler<HTMLInputElement> = e => {
        if (!/^(\d+(,|\s))*.?(\d*)$/.test(e.data)) {
            e.preventDefault();
        }
    };
```

Reply:
Epic, I didn't realize there is a onBeforeInput event 😅  b9d34d720

Reply:
🚀